### DISCUSSION TITLE
How to connect via ssh command

### DISCUSSION THREAD

**Discussion by FallenStark** [gravitational/teleport#29807] [Q&A] [answered]
Hi all, I did a teleport setup, installed agents on my servers, but we don't want to use the "tsh" tool to connect to the servers via SSH.

Could you please tell me how to implement SSH connection via the built-in "ssh" command in Linux, thanks!

I will be very grateful for more detailed answer.

**Comment by webvictim** [comment]
The simplest answer is:
- log in with `tsh login`
- run `tsh config >> ~/.ssh/config` to add a Teleport SSH config to your local SSH config file
- run `ssh user@teleport-hostname` to connect to a Teleport host via the proxy using the SSH client

`tsh help config` should give a little more information. You can also look at the output of `tsh config` to see what it's doing. You may need to tweak the hostname matching a little to make it work the way you want it to.

**Reply by FallenStark** [reply]
![2](https://github.com/gravitational/teleport/assets/90399485/895a4cf0-5d38-497d-ae55-b0f9afdd1754)


I enter the username as in the URL (web) teleport @ and the domain name of the teleport server

**Comment by FallenStark** [comment]
Thank you so much for the reply, I am getting this error:
![1](https://github.com/gravitational/teleport/assets/90399485/4c3e371e-27bc-44a5-9bea-0704f76c5f96)

I specified ssh loginteleport@teleport-hostname(domain name)

**Reply by webvictim** [reply]
Can you share the full output of `ssh -vv login@hostname`? You can redact the domain name or IP addresses, but please leave other information intact as it's important to understand what SSH is doing when attempting to connect to the host.

Your `login` used here should be one of the logins allowed on the server in a Teleport role, i.e. the `Logins:` output of `tsh status`.

**Reply by FallenStark** [reply]
Thanks so much for the reply! Here are the screenshots:
![3](https://github.com/gravitational/teleport/assets/90399485/9ab1cc41-cf58-4f68-b03b-b1bc86410435)
![4](https://github.com/gravitational/teleport/assets/90399485/66e7fc21-bf67-42dc-a23d-9eb9a9874a8f)
![5](https://github.com/gravitational/teleport/assets/90399485/dca797ce-85e5-4e02-89cc-c48d01edf614)

**Reply by webvictim** [reply]
I think the issue here is that you are attempting to SSH into to the Teleport proxy itself (`teleport.xxx.com`) rather than a node behind the proxy. You can tell because SSH is attempting to connect over regular SSH (i.e. port 22) rather than using the `ProxyCommand` which `tsh config` wrote into the `.ssh/config` file.

What nodes do you see in the output of `tsh ls`? Can you try SSHing to one of those? Something like `ssh -vv login@servername.teleport.xxx.com` if you see `servername` in the output of `tsh ls`.

**Reply by FallenStark** [reply]
Thanks a lot for the reply! For some reason it resolves to a local IP (wrong as in AWS Amazon in Hostname).
![6](https://github.com/gravitational/teleport/assets/90399485/4224ccaa-ea59-48cf-b27e-bfd47f4c7de5)

**Reply by webvictim** [reply]
It appears that the reason is that your `~/.ssh/config` file contains an existing entry for `dolphin-anty-site-prod` on line 1 as per the 3rd line down in the logs.

Try `ssh -vv root@dolphin-anty-site-prod.teleport.xxx.com` (replace `xxx` with your domain)

You could also try connecting to your Teleport proxy using `ssh -vv root@ip-10-90-2-72.teleport.xxx.com`

**Reply by FallenStark** [reply]
![7](https://github.com/gravitational/teleport/assets/90399485/32450655-7973-4c71-994a-3e342991c2ee)
![8](https://github.com/gravitational/teleport/assets/90399485/f195d5fe-5287-4587-9200-8315f6ba0d01)
Unfortunately it doesn't work that way :(

**Reply by webvictim** [reply]
Ah - I just realised your cluster is called `teleport-dolphin-main` rather than `teleport.xxx.com`.

Try `ssh -vv ubuntu@dolphin-anty-site-prod.teleport-dolphin-main` or `ssh -vv ubuntu@ip-10-90-2-72.teleport.dolphin-main`

**Reply by FallenStark** [reply]
![9](https://github.com/gravitational/teleport/assets/90399485/84adc2d6-d4dc-470e-9ec1-335a1cc6ecad)
![10](https://github.com/gravitational/teleport/assets/90399485/c243f6e8-0cb1-41ba-9b49-4eb2e44252be)
![11](https://github.com/gravitational/teleport/assets/90399485/73e6d1e1-f7b9-45ec-954b-e1666f8e9c18)

That's another mistake

**Reply by webvictim** [reply]
OK, we're definitely getting closer. Try this:
```
ssh -o "HostKeyAlgorithms=rsa-sha2-512-cert-v01@openssh.com,rsa-sha2-256-cert-v01@openssh.com,ssh-rsa-cert-v01@openssh.com" -vv ubuntu@dolphin-anty-site-prod.teleport-dolphin-main
```

**Reply by FallenStark** [reply]
![12](https://github.com/gravitational/teleport/assets/90399485/335fedd4-e39b-4bc6-b314-c7f8eb7bddeb)
![13](https://github.com/gravitational/teleport/assets/90399485/23423d69-a49a-41dc-a44b-50ac2a33ab83)

**Reply by webvictim** [reply]
Closer still!

`ssh -o "HostKeyAlgorithms=rsa-sha2-512-cert-v01@openssh.com,rsa-sha2-256-cert-v01@openssh.com,ssh-rsa-cert-v01@openssh.com" -o "PubkeyAcceptedAlgorithms=rsa-sha2-512-cert-v01@openssh.com,rsa-sha2-256-cert-v01@openssh.com,ssh-rsa-cert-v01@openssh.com" -vv ubuntu@dolphin-anty-site-prod.teleport-dolphin-main`

**Reply by FallenStark** [reply]
![14](https://github.com/gravitational/teleport/assets/90399485/6ee10625-22a0-4037-93f1-920e0d7ef4bd)
![15](https://github.com/gravitational/teleport/assets/90399485/c2bc82de-1d64-478c-abed-8db88e97625f)

**Reply by webvictim** [reply, marked as answer]
Oh, that's my bad:

`ssh -o "HostKeyAlgorithms=rsa-sha2-512-cert-v01@openssh.com,rsa-sha2-256-cert-v01@openssh.com,ssh-rsa-cert-v01@openssh.com" -o "PubkeyAcceptedAlgorithms=ssh-rsa-cert-v01@openssh.com,ssh-rsa" -vv ubuntu@dolphin-anty-site-prod.teleport-dolphin-main`

**Reply by FallenStark** [reply]
![image](https://github.com/gravitational/teleport/assets/90399485/f614a85d-e08d-4402-989c-f0cee8f3cba8)

It's working! Thanks a lot :)

**Reply by webvictim** [reply]
Nice! Sorry about the confusion - you can probably tidy a lot of this up by specifying the options directly in the config file rather than on the command line in future.

**Comment by ghost** [comment]
is there any way to access a virtual machine that is in teleport using go?
